### PR TITLE
Fix CVE-2017-18350

### PR DESCRIPTION
**Pull request by ebfull**
Adapted from bitcoin/bitcoin#11397 by Wladimir J. van der Laan.

### CODE REVIEW

**Comment by str4d** [comment]
CI is running here: https://ci.z.cash/#/builders/3/builds/768

**Comment by daira** [comment]
As a security fix this needs three reviewers.

**Comment by zebambam** [comment]
Looks like a good fix to me.

File: src/netbase.cpp [hunk 1]
```diff
@@ -246,7 +246,7 @@ struct timeval MillisToTimeval(int64_t nTimeout)
  *
  * @note This function requires that hSocket is in non-blocking mode.
  */
-bool static InterruptibleRecv(char* data, size_t len, int timeout, SOCKET& hSocket)
+bool static InterruptibleRecv(uint8_t* data, size_t len, int timeout, SOCKET& hSocket)
```
Comment (Eirik0) [inline src/netbase.cpp]: In the upstream PR `data` is cast back to a `char*` in the body of this function:
```ssize_t ret = recv(hSocket, (char*)data, len, 0); // Optimistically try the recv first```
Why is that not done here?

Comment (ebfull) [reply src/netbase.cpp]: It's unnecessary, as @daira could explain better. I think it was done in the upstream PR to reduce the worry about the PR introducing a bug and to perform the fix more covertly, or maybe due to a misunderstanding.

File: src/netbase.cpp [hunk 2]
```diff
@@ -408,7 +408,7 @@ static bool Socks5(const std::string& strDest, int port, const ProxyCredentials
                 CloseSocket(hSocket);
                 return error("Error reading from proxy");
             }
-            int nRecv = pchRet3[0];
+            size_t nRecv = pchRet3[0];
```
Comment (Eirik0) [inline src/netbase.cpp]: This look good, but I do not see this in the upstream PR. I am curious the upstream PR does not include this.

Comment (ebfull) [reply src/netbase.cpp]: It's probably not in the upstream PR to reduce the risk that someone would identify it as a covert fix of a security issue.

**Comment by milesmanley** [comment]
👍 Thanks for the quick fix and great work as always

**Comment by str4d** [comment]
CI passed on two of our four supported builders, with the other two suffering from EC2 timeouts. Given that we'll be doing another CI run on the release PR, I'm happy to move forward.

**Comment by cronicc** [comment]
Thanks for the quick reaction, greatly appreciated.

**Pull request merged**